samples/bluetooth/hci_pwr_ctrl/src/main.c

```diff
@@ -0,0 +1,318 @@
+/* main.c - Application main entry point */
+
+/*
+ * Copyright (c) 2019 Andrei Stoica
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <sys/printk.h>
+#include <sys/util.h>
+#include <sys/byteorder.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/hci.h>
+#include <bluetooth/hci_vs.h>
+
+#include <bluetooth/conn.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/services/hrs.h>
+
+static struct bt_conn *default_conn;
+static u16_t default_conn_handle;
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA_BYTES(BT_DATA_UUID16_ALL, 0x0d, 0x18),
+};
+
+#define DEVICE_NAME CONFIG_BT_DEVICE_NAME
+#define DEVICE_NAME_LEN (sizeof(DEVICE_NAME) - 1)
+#define DEVICE_BEACON_TXPOWER_NUM  8
+
+static struct k_thread pwr_thread_data;
+static K_THREAD_STACK_DEFINE(pwr_thread_stack, 320);
+
+static const s8_t txp[DEVICE_BEACON_TXPOWER_NUM] = {4, 0, -3, -8,
+						    -15, -18, -23, -30};
+static const struct bt_le_adv_param *param = BT_LE_ADV_PARAM
+	(BT_LE_ADV_OPT_CONNECTABLE | BT_LE_ADV_OPT_USE_NAME,
+	0x0020, 0x0020);
+
+static void read_conn_rssi(u16_t handle, s8_t *rssi)
+{
+	struct bt_hci_cp_read_rssi *cp;
+	struct bt_hci_rp_read_rssi *rp;
+	struct net_buf *buf, *rsp;
+	int err;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_READ_RSSI, sizeof(*cp));
+	if (!buf) {
+		printk("Unable to allocate command buffer\n");
+		return;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	cp->handle = sys_cpu_to_le16(handle);
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_READ_RSSI, buf, &rsp);
+	if (err) {
+		printk("RSSI HCI err: %d\n", err);
```

Comment:
This is not HCI error. Also `bt_hci_cmd_send_sync` can fail before receiving the rsp.
So we can't safely use rsp if err is received.
I would suggest to do this:
```
struct net_buf *buf, *rsp = NULL;
...
if (err) {
	u8_t reason = rsp ? ((struct bt_hci_rp_read_rssi *)rsp->data)->status : 0;
	printk("Read RSSI err: %d\n reason 0x%02x", err, status);
	return;
}
```
Needs to be handled for the other cases where `rsp` is used in this way as well.

Reply:
Thanks for pointing it out! I see why now. I saw now that `bt_hci_cmd_send_sync` may fail without allocation of `*rsp`. 

I implemented the changes throughout in all the functions handling the HCI calls. Here the changes for the `read_conn_rssi` - [pointer initialization](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/c8f060e0436bb073728d128cfba438c930a1f323#diff-b9bd630d9245e9d01d535cbaa393a233R47) and [exit logic](https://github.com/zephyrproject-rtos/zephyr/pull/19779/commits/c8f060e0436bb073728d128cfba438c930a1f323#diff-b9bd630d9245e9d01d535cbaa393a233R63-R68)